dll/win32/shell32/shellrecyclebin/recyclebin_v5_enumerator.cpp

```diff
@@ -0,0 +1,462 @@
+    if (GetFileAttributesW(szFullName) == INVALID_FILE_ATTRIBUTES)
+        return E_FAIL;
+
+    LPWSTR pszFullName;
+    HRESULT hr = SHStrDup(szFullName, &pszFullName);
+    if (FAILED(hr))
+        return hr;
+
+    RecycleBin5File *pThis = new RecycleBin5File(prb, Folder, pDeletedFile, pszFullName);
+    if (!pThis)
+        return E_OUTOFMEMORY;
+
+    *ppFile = static_cast<IRecycleBinFile *>(pThis);
+    return S_OK;
+}
+
+class RecycleBin5Enum : public IRecycleBinEnumList
+{
+public:
+    RecycleBin5Enum(
+        IN IRecycleBin5 *prb,
+        IN HANDLE hInfo,
+        IN OUT INFO2_HEADER *pInfo,
+        IN OUT LPWSTR pszPrefix);
+    virtual ~RecycleBin5Enum();
+
+    /* IUnknown methods */
+    STDMETHODIMP QueryInterface(REFIID riid, void **ppvObject) override;
+    STDMETHODIMP_(ULONG) AddRef() override;
+    STDMETHODIMP_(ULONG) Release() override;
+
+    /* IRecycleBinEnumList methods */
+    STDMETHODIMP Next(DWORD celt, IRecycleBinFile **rgelt, DWORD *pceltFetched) override;
+    STDMETHODIMP Skip(DWORD celt) override;
+    STDMETHODIMP Reset() override;
+
+protected:
+    LONG m_ref;
+    IRecycleBin5 *m_recycleBin;
+    HANDLE m_hInfo;
+    INFO2_HEADER *m_pInfo;
+    DWORD m_dwCurrent;
+    LPWSTR m_pszPrefix;
+};
+
+STDMETHODIMP RecycleBin5Enum::QueryInterface(REFIID riid, void **ppvObject)
+{
+    TRACE("(%p, %s, %p)\n", this, debugstr_guid(&riid), ppvObject);
+
+    if (!ppvObject)
+        return E_POINTER;
+
+    if (IsEqualIID(riid, IID_IUnknown) || IsEqualIID(riid, IID_IRecycleBinEnumList))
+        *ppvObject = static_cast<IRecycleBinEnumList *>(this);
+    else
+    {
+        *ppvObject = NULL;
+        return E_NOINTERFACE;
+    }
+
+    AddRef();
+    return S_OK;
+}
+
+STDMETHODIMP_(ULONG) RecycleBin5Enum::AddRef()
+{
+    TRACE("(%p)\n", this);
+    return InterlockedIncrement(&m_ref);
+}
+
+RecycleBin5Enum::~RecycleBin5Enum()
+{
+    TRACE("(%p)\n", this);
+
+    m_recycleBin->OnClosing(this);
+    UnmapViewOfFile(m_pInfo);
+    m_recycleBin->Release();
+    SHFree(m_pszPrefix);
+}
+
+STDMETHODIMP_(ULONG) RecycleBin5Enum::Release()
+{
+    TRACE("(%p)\n", this);
+
+    ULONG refCount = InterlockedDecrement(&m_ref);
+    if (refCount == 0)
+        delete this;
+    return refCount;
+}
+
+STDMETHODIMP RecycleBin5Enum::Next(DWORD celt, IRecycleBinFile **rgelt, DWORD *pceltFetched)
+{
+    HRESULT hr;
+
+    TRACE("(%p, %u, %p, %p)\n", this, celt, rgelt, pceltFetched);
+
+    if (!rgelt)
+        return E_POINTER;
+    if (!pceltFetched && celt > 1)
+        return E_INVALIDARG;
+
+    ULARGE_INTEGER FileSize;
+    FileSize.u.LowPart = GetFileSize(m_hInfo, &FileSize.u.HighPart);
+    if (FileSize.u.LowPart == 0)
+        return HRESULT_FROM_WIN32(GetLastError());
+
+    DWORD dwEntries =
+        (DWORD)((FileSize.QuadPart - sizeof(INFO2_HEADER)) / sizeof(DELETED_FILE_RECORD));
+
+    DWORD iEntry = m_dwCurrent, fetched = 0;
+    DELETED_FILE_RECORD *pDeletedFile = (DELETED_FILE_RECORD *)(m_pInfo + 1) + iEntry;
```

Comment:
All the `DELETED_FILE_RECORD *` --> `PDELETED_FILE_RECORD`

Reply:
Fixed.